### DISCUSSION TITLE
Plastic parameters that depend on space and time

### DISCUSSION THREAD

**Discussion by Hadrien89** [idaholab/moose#21063] [Q&A Modules: Solid mechanics] [answered]
Hello,

I’m trying to run slope stability simulations in tensor mechanics. In particular, I’m using the shear strength reduction method on a heterogenous material. For this method, I would need to have the friction angle and the cohesion in the Mohr-Coulomb model that both depend on space (given by a csv file) and time (all the Mohr-Coulomb parameters are reduced by the same factor at every time step).

I have tried to follow the solution used in Porous Flow to produce a heterogeneous reservoir:
https://mooseframework.inl.gov/modules/porous_flow/heterogeneous_models.html 

This solution does not to be directly applicable in my case as the materials properties in this example that change with space (permeability and porosity) are declared as Auxvariables and are used in a Material. However, it seems that I cannot used Auxvariables in a UserObject, which is the current way to define the Mohr-Coulomb properties. I get the following error when compiling if I try to create a Userobject that could use the Auxvariable of the friction angle:

error: use of undeclared identifier 'coupledValue'; did you mean 'coupledScalar'?

It seems that only Scalar variables can be used in a UserObject. 
Do you have some ideas how I could have those parameters depending on space and time (without changing the whole return map algorithm)?

Thanks for your help!

**Comment by GiudGiud** [comment, marked as answer]
Hello

You need your class to be using the `Coupleable` interface to be able to use the `coupledValue` routine. This interface is inherited by Element/Side/Interface/etcUserObject, not by UserObject the parent class of all of these.

If your UO fits in one of these categories, please inherit those instead. 

For parameters depending on space and time, you can use a Function if they are known at the time you are writing your input file. If they are to be computed during the simulation you can use:
- a material property
- an auxiliary variable if you dont need to take derivatives into account / the terms are not part of forming a Jacobian with automatic differentiation. OR if you need tight control on when to update the quantity (linear / nonlinear iterations / once a timestep etc)

Guillaume

**Reply by Hadrien89** [reply]
Hi Guillaume,

Thanks a lot for your quick reply!

Regarding the solution you propose for the parameters that depend on space and time, it is what I’m trying to do. I’m using a function that is changing an AuxVariable for each parameter, but I’m stuck at using this AuxVar in a UO…

To be more precise, the UO I have created to have the plastic properties coming from an AuxVariable is a modified copy of the UO “TensorMechanicsHardeningConstant” (which takes constant parameters as input). The parent class of all UO that defines plastic parameters is “TensorMechanicsHardeningModel”, which inherits from GeneralUserObject and, as you suggested, it seems to be the issue here. 
I have tried to have “TensorMechanicsHardeningModel” inherit from Element/Side/Interface/etcUserObject instead of GeneralUserObject, but it is creating 20 errors when compiling. 
I also tried to #include different files (like "MooseVariable.h" and others) directly in the UO I have created, but it didn’t change anything.
I apologise my knowledge of C++ is quite limited...

Hadrien

**Reply by GiudGiud** [reply]
Hello

The easiest way might just be to add the Coupleable interface as a parent class for your class then. 
Can you try that?
SideUserObject.h/C or any of the others I mentioned will show the two lines you need to add to do this. 

Guillaume

**Reply by GiudGiud** [reply]
@Hadrien89 any luck with this?
